src/v/cluster/scheduling/leader_balancer_constraints.h

```diff
@@ -0,0 +1,337 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "cluster/partition_leaders_table.h"
+#include "cluster/scheduling/leader_balancer_types.h"
+#include "cluster/topic_table.h"
+#include "cluster/types.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "raft/types.h"
+
+#include <seastar/core/metrics.hh>
+#include <seastar/core/sstring.hh>
+
+#include <absl/container/btree_map.h>
+#include <absl/container/flat_hash_map.h>
+#include <absl/container/flat_hash_set.h>
+#include <boost/range/adaptor/reversed.hpp>
+
+#include <cstdint>
+#include <functional>
+#include <limits>
+#include <numeric>
+#include <optional>
+#include <unordered_map>
+
+namespace cluster::leader_balancer_types {
+
+class shard_index final : public index {
+public:
+    explicit shard_index(index_type index)
+      : _index(std::move(index)) {}
+
+    shard_index(shard_index&& si) noexcept
+      : _index(std::move(si._index)) {}
+    shard_index(const shard_index& si) noexcept
+      : _index(si._index) {}
+
+    shard_index& operator=(shard_index&&) = default;
+    shard_index& operator=(const shard_index&) = default;
+
+    ~shard_index() override = default;
+
+    void update_index(const reassignment& r) override {
+        auto replicas = std::move(_index[r.from][r.group]);
+        _index[r.to].try_emplace(r.group, std::move(replicas));
+        _index[r.from].erase(r.group);
+    }
+
+    const index_type& shards() const { return _index; }
+
+private:
+    index_type _index;
+};
+
+/*
+ * Tracks nodes and groups being `muted` by the leader balancer.
+ * Reassignments to and from muted nodes and groups should be avoided.
+ */
+class muted_index final : public index {
+public:
+    muted_index(
+      absl::flat_hash_set<model::node_id> muted_nodes,
+      absl::flat_hash_set<raft::group_id> muted_groups)
+      : _muted_nodes(std::move(muted_nodes))
+      , _muted_groups(std::move(muted_groups)) {}
+
+    muted_index(muted_index&& mi) noexcept
+      : _muted_nodes(std::move(mi._muted_nodes))
+      , _muted_groups(std::move(mi._muted_groups)) {}
```

Comment:
seems like the default implementation would work?

Reply:
Will switch to default implementations.